lib/elixir/pages/Library Guidelines.md

```diff
@@ -0,0 +1,214 @@
+
+```elixir
+def split(string, opts \\ []) when is_binary(string) and is_list(opts) do
+  parts = Keyword.get(opts, :parts, 2)
+  String.split(string, "-", parts: parts)
+end
+```
+
+In case you need to configure a process, the options should be passed when starting that process. 
+
+The application environment should be reserved only for configurations that are truly global, for example, to control your application boot process and its supervision tree.
+
+### Avoid `use` when an `import` is enough
+
+A library should not provide `use MyLib` functionality if all that `use MyLib` does is to `import`/`require` the module itself. For example, this is an anti-pattern:
+
+```elixir
+defmodule MyLib do
+  defmacro __using__(_) do
+    quote do
+      import MyLib # or require MyLib
+    end
+  end
+
+  def some_fun(arg1, arg2) do
+    ...
+  end
+end
+```
+
+The reason why defing the `__using__` macro above should be avoided is because when a developer writes
+
+```elixir
+defmodule MyApp do
+  use MyLib
+end
+```
+
+it allows `use MyLib` to run *any* code into the `MyApp` module. For someone reading the code, it is impossible to assess the impact that `use MyLib` has in a module.
+
+On the other hand, the following code is much clearer:
+
+```elixir
+defmodule MyApp do
+  import MyLib
+end
+```
+
+The code above says we are only bringing in the functions from `MyLib` so we can invoke `some_fun(arg1, arg2)` directly without the `MyLib.` prefix. Even more important, `import MyLib` says that we have an option to not `import MyLib` at all as we can simply invoke the function as `MyLib.some_fun(arg1, arg2)`.
+
+While there are many situations where using a module is required, `use` should be skipped when all it does is to `import` or `require` a module. `import` and `require` are a simpler contract that gives users of your library more control on how to design and maintain their code. 
+
+### Avoid macros
+
+Although the previous section could be resumed as "avoid macros", both topics are important enough to deserve their own section.
+
+To quote [the official guide on Macros](https://elixir-lang.org/getting-started/meta/macros.html):
+
+> Even though Elixir attempts its best to provide a safe environment for macros, the major responsibility of writing clean code with macros falls on developers. Macros are harder to write than ordinary Elixir functions and it’s considered to be bad style to use them when they’re not necessary. So write macros responsibly.
+>
+> Elixir already provides mechanisms to write your everyday code in a simple and readable fashion by using its data structures and functions. Macros should only be used as a last resort. Remember that **explicit is better than implicit**. **Clear code is better than concise code**.
+
+When you absolutely have to use macro, make sure that a macro is not the only way the user can interface with your library and keep the amount of code generated by a macro to a minimum.
```

Comment:
I think an example could be useful here because it's pretty abstract. One example we could use is how Ecto moved from validation macros to pure data structure changesets.

Reply:
The Ecto example is too specific. I would like something more contained but nothing comes to mind.

Reply:
When you absolutely have to use **a** macro,
Or
When you absolutely have to use **macros**,

Reply:
The example that sprung to my mind was how benchfella uses macros `bench "bla" do ... end` and a custom `mix bench` vs. benchee `Benchee.benchmark("bla", fn -> ... end)` and run as a normal script.

Avoiding needless macros was one of the major reasons I started writing benchee.

Example might be too specific still.

Reply:
What about the Logger's `info/2`, `debug/2`, `log/3` etc macros vs the `bare_log/3` function?

Reply:
@tompave jackpot! I will go with that. :)